File: src/main/csharp/SunliteSharp/Core/Compiler/Scanner.cs
using System.Diagnostics.CodeAnalysis;
using System.Text;
using SunliteSharp.Core.Enum;
using static SunliteSharp.Core.Enum.TokenType;

namespace SunliteSharp.Core.Compiler;

[SuppressMessage("ReSharper", "InconsistentNaming")]
public class Scanner(string source, Sunlite sunlite)
{
    private string CurrentFile = "";
    
    private readonly List<Token> Tokens = [];

    private int Start = 0;
    private int Current = 0;
    private int Line = 1;
    private int TotalChars = 0;
    private int LineStart = 0;
    private int LineCurrent = 0;

    private static readonly Dictionary<string, TokenType> Keywords = new()
    {
        { "import", Import },
        { "and", And },
        { "or", Or },
        { "true", True },
        { "false", False },
        { "nil", Nil },
        { "if", If },
        { "else", Else },
        { "for", For },
        { "while", While },
        { "break", Break },
        { "continue", Continue },
        { "try", Try },
        { "catch", Catch },
        { "throw", Throw },
        { "fun", Fun },
        { "class", Class },
        { "interface", Interface },
        { "extends", Extends },
        { "implements", Implements },
        { "init", Init },
        { "this", This },
        { "super", Super },
        { "return", Return },
        { "var", Var },
        { "val", Val },
        { "static", Static },
        { "native", Native },
        { "is", Is },
        { "isnt", IsNot },
        { "as", As },
        { "Any", TypeAny },
        { "Boolean", TypeBoolean },
        { "Number", TypeNumber },
        { "String", TypeString },
        { "Array", TypeArray },
        { "Table", TypeTable },
        { "Generic", TypeGeneric },
        { "Class", TypeClass },
        { "Function", TypeFunction },
        { "Nil", TypeNil },
    };

    public List<Token> ScanTokens(string path)
    {
        CurrentFile = path;
        
        while (!IsAtEnd()) {
            // We are at the beginning of the next lexeme.
            Start = Current;
            LineStart = LineCurrent;
            ScanToken();
        }
        
        Tokens.Add(new Token(Eof, "", null, Line, CurrentFile, new Token.Position(0, 0)));
        return Tokens;
    }

    private bool IsAtEnd()
    {
        return Current >= source.Length;
    }
    
    private void ScanToken()
    {
        var c = Advance();
        switch (c)
        {
            case '(': AddToken(LeftParen); break;
            case ')': AddToken(RightParen); break;
            case '{': AddToken(LeftBrace); break;
            case '}': AddToken(RightBrace); break;
            case '[': AddToken(LeftBracket); break;
            case ']': AddToken(RightBracket); break;
            case ',': AddToken(Comma); break;
            case '.': AddToken(Dot); break;
            case '-': AddToken(Match('=') ? MinusEqual : Minus); break;
            case '+': AddToken(Match('=') ? PlusEqual : Plus); break;
            case ';': AddToken(Semicolon); break;
            case '*': AddToken(Star); break;
            case ':': AddToken(Colon); break;
            case '|': AddToken(Pipe); break;
            case '?': AddToken(Question); break;
            case '!': AddToken(Match('=') ? BangEqual : Bang); break;
            case '=': AddToken(Match('=') ? EqualEqual : Equal); break;
            case '<': AddToken(Match('=') ? LessEqual : Less); break;
            case '>': AddToken(Match('=') ? GreaterEqual : Greater); break;
            case '/':
                if (Match('/'))
                {
                    while (Peek() != '\n' && !IsAtEnd()) Advance();
                }
                else if (Match('*'))
                {
                    while (Peek() != '*' && PeekNext() != '/' && !IsAtEnd()) Advance();
                    if (!IsAtEnd() && Peek() == '*' && PeekNext() == '/')
                    {
                        Advance();
                        Advance();
                    }
                }
                else
                {
                    AddToken(Slash);
                }
                break;
            case ' ':
            case '\r':
            case '\t':
                break;
            case '\n':
                LineStart = 0;
                LineCurrent = 0;
                Line++;
                break;
            case '"':
                String();
                break;
            default:
                if (char.IsDigit(c))
                {
                    Number();
                }
                else if (char.IsAsciiLetter(c) || c == '_')
                {
                    Identifier();
                }
                else
                {
                    sunlite.Error($"Unexpected character '{c}'", Line, CurrentFile);
                }
                break;
        }
    }

    private void AddToken(TokenType type, object? literal = null) 
    {
        var text = source.Substring(Start, Current - Start);
        Tokens.Add(new Token(type, text, literal, Line, CurrentFile, new Token.Position(LineStart, LineCurrent)));
    }

    private char Advance()
    {
        TotalChars++;
        LineCurrent++;
        return source[Current++];
    }

    private char Peek()
    {
        return IsAtEnd() ? '\u0000' : source[Current];
    }
    
    private char PeekNext()
    {
        return Current + 1 >= source.Length ? '\u0000' : source[Current + 1];
    }

    private bool Match(char expected)
    {
        if (IsAtEnd()) return false;
        if (source[Current] != expected) return false;
        Current++;
        TotalChars++;
        LineCurrent++;
        return true;
    }

    private void String()
    {
        var sb = new StringBuilder();
        while (Peek() != '"' && !IsAtEnd())
        {
            if(Peek() == '\n') Line++;
            
            //escape sequences
            if (Peek() == '\\')
            {
                Advance();
                var ch = Peek();
                char formatChar = ' ';
                switch (ch)
                {
                    case 'n':
                        formatChar = '\n';
                        break;
                    case '\\' :
                        formatChar = '\\';
                        break;
                    case '"':
                        formatChar = '"';
                        break;
                    default:
                        sunlite.Error("Unknown escape sequence.", Line, CurrentFile);
                        break;
                }
                Advance();
                sb.Append(formatChar);
            }

            //interpolation
            if (Peek() == '$' && PeekNext() == '{')
            {
                AddToken(TokenType.String, sb.ToString());
                Advance();
                Advance();
                
                Tokens.Add(new Token(Plus, "+", null, Line, CurrentFile, new Token.Position(LineStart, LineCurrent)));
                
                var iSb = new StringBuilder();
                while (Peek() != '}' && !IsAtEnd())
                {
                    iSb.Append(Peek());
                    Advance();
                }
                
                var innerScanner = new Scanner(iSb.ToString(), sunlite);
                List<Token> interpolatedTokens = innerScanner.ScanTokens(CurrentFile);
                interpolatedTokens = interpolatedTokens[..^1];
                Console.WriteLine(interpolatedTokens);
                
                Tokens.AddRange(interpolatedTokens);
                
                Tokens.Add(new Token(Plus, "+", null, Line, CurrentFile, new Token.Position(LineStart, LineCurrent)));

                if (IsAtEnd())
                {
                    sunlite.Error("Expected '}' after string interpolated expression.", Line, CurrentFile);
                    return;
                }
                Advance();
                sb.Clear();
            }

            sb.Append(Peek());
            Advance();
        }

        if (IsAtEnd())
        {
            sunlite.Error("Unterminated string.", Line, CurrentFile);
        }

        Advance();
        
        AddToken(TokenType.String, sb.ToString());
    }

    private void Identifier()
    {
        while (char.IsLetterOrDigit(Peek()) || Peek() == '_')
        {
            Advance();
        }
        
        var text = source.Substring(Start, Current - Start);
        var type = Keywords.GetValueOrDefault(text, TokenType.Identifier);
        AddToken(type);
    }

    private void Number()
    {
        while (char.IsDigit(Peek()))
        {
            Advance();
        }

        if (Peek() == '.' && char.IsDigit(PeekNext()))
        {
            Advance();
            while (char.IsDigit(Peek()))
            {
                Advance();
            }
        }
        
        AddToken(TokenType.Number, double.Parse(source.Substring(Start, Current - Start)));
    }
}

File: src/main/kotlin/sunsetsatellite/lang/sunlite/Scanner.kt
package sunsetsatellite.lang.sunlite

import sunsetsatellite.lang.sunlite.TokenType.*


class Scanner(private val source: String, val sunlite: Sunlite) {
	private val tokens: ArrayList<Token> = ArrayList()
	private var start = 0
	private var current = 0
	private var line = 1
	private var totalChars = 0
	private var lineStart = 0
	private var lineCurrent = 0

	var currentFile: String? = null

	companion object {
		private val keywords: MutableMap<String, TokenType> = mutableMapOf()

		init {
			keywords["and"] = AND
			keywords["class"] = CLASS
			keywords["else"] = ELSE
			keywords["false"] = FALSE
			keywords["for"] = FOR
			keywords["fun"] = FUN
			keywords["init"] = INIT
			keywords["if"] = IF
			keywords["nil"] = NIL
			keywords["or"] = OR
			//keywords["print"] = PRINT
			keywords["return"] = RETURN
			keywords["super"] = SUPER
			keywords["this"] = THIS
			keywords["true"] = TRUE
			keywords["var"] = VAR
			keywords["val"] = VAL
			keywords["while"] = WHILE
			keywords["break"] = BREAK
			keywords["continue"] = CONTINUE
			keywords["static"] = STATIC
			keywords["native"] = NATIVE
			keywords["interface"] = INTERFACE
			keywords["is"] = IS
			keywords["isnt"] = IS_NOT
			keywords["import"] = IMPORT
			keywords["dynamic"] = DYNAMIC
			keywords["Any"] = TYPE_ANY
			keywords["String"] = TYPE_STRING
			keywords["Number"] = TYPE_NUMBER
			keywords["Boolean"] = TYPE_BOOLEAN
			keywords["Function"] = TYPE_FUNCTION
			keywords["Array"] = TYPE_ARRAY
			keywords["Table"] = TYPE_TABLE
			keywords["Generic"] = TYPE_GENERIC
			keywords["Class"] = TYPE_CLASS
			keywords["Nil"] = TYPE_NIL
			keywords["as"] = AS
			keywords["extends"] = EXTENDS
			keywords["implements"] = IMPLEMENTS
			keywords["try"] = TRY
			keywords["catch"] = CATCH
			keywords["throw"] = THROW
		}
	}


	fun scanTokens(path: String?): List<Token> {
		currentFile = path

		while (!isAtEnd()) {
			// We are at the beginning of the next lexeme.
			start = current
			lineStart = lineCurrent
			scanToken()
		}

		tokens.add(Token(EOF, "", null, line, currentFile, Token.Position(-1,-1)))
		return tokens
	}

	private fun isAtEnd(): Boolean {
		return current >= source.length
	}

	private fun scanToken() {
		when (val c = advance()) {
			'(' -> addToken(LEFT_PAREN)
			')' -> addToken(RIGHT_PAREN)
			'{' -> addToken(LEFT_BRACE)
			'}' -> addToken(RIGHT_BRACE)
			'[' -> addToken(LEFT_BRACKET)
			']' -> addToken(RIGHT_BRACKET)
			',' -> addToken(COMMA)
			'.' -> addToken(DOT)
			'-' -> addToken(if(match('=')) MINUS_EQUAL else MINUS)
			'+' -> addToken(if(match('=')) PLUS_EQUAL else PLUS)
			';' -> addToken(SEMICOLON)
			'*' -> addToken(STAR)
			':' -> addToken(COLON)
			'|' -> addToken(PIPE)
			'?' -> addToken(QUESTION)
			'!' -> addToken(if(match('=')) BANG_EQUAL else BANG)
			'=' -> addToken(if(match('=')) EQUAL_EQUAL else EQUAL)
			'<' -> addToken(if(match('=')) LESS_EQUAL else LESS)
			'>' -> addToken(if(match('=')) GREATER_EQUAL else GREATER)
			'/' -> {
				if(match('/')){
					while (peek() != '\n' && !isAtEnd()) advance()
				} else if(match('*')){
					while (peek() != '*' && peekNext() != '/' && !isAtEnd()) advance()
					if(!isAtEnd() && peek() == '*' && peekNext() == '/') advance(); advance()
				}
				else {
					addToken(SLASH)
				}
			}
			' ', '\r', '\t' -> {}
			'\n' -> {
				lineStart = 0
				lineCurrent = 0
				line++
			}
			'"' -> string()
			in '0'..'9' -> number()

			else -> {

				if(isAlpha(c)) {
					identifier()
					return
				}

				sunlite.error(line, "Unexpected character '$c'")
			}
		}
	}

	private fun match(expected: Char): Boolean {
		if (isAtEnd()) return false
		if (source[current] != expected) return false

		current++
		totalChars++
		lineCurrent++
		return true
	}

	private fun peek(): Char {
		if (isAtEnd()) return '\u0000'
		return source[current]
	}

	private fun peekNext(): Char {
		if (current + 1 >= source.length) return '\u0000'
		return source[current + 1]
	}


	private fun advance(): Char {
		totalChars++
		lineCurrent++
		return source[current++]
	}

	private fun addToken(type: TokenType) {
		addToken(type, null)
	}

	private fun addToken(type: TokenType, literal: Any?) {
		val text = source.substring(start, current)
		tokens.add(Token(type, text, literal, line, currentFile, Token.Position(lineStart,lineCurrent)))
	}

	private fun string() {
		val sb = StringBuilder()
		while (peek() != '"' && !isAtEnd()) {
			if (peek() == '\n') line++

			// Escape sequences
			if(peek() == '\\'){
				advance()
				when(advance()){
					'n' -> sb.append('\n')
					'r' -> sb.append('\r')
					't' -> sb.append('\t')
					'\\' -> sb.append('\\')
					'"' -> sb.append('\"')
					else -> {
						sunlite.error(line, "Invalid escape sequence.")
						return
					}
				}
			}

			if(peek() == '$' && peekNext() == '{'){
				addToken(STRING, sb.toString())
				advance()
				advance()

				tokens.add(Token(PLUS, "+", null, line, currentFile, Token.Position(lineStart,lineCurrent)))

				val isb = StringBuilder()
				while(peek() != '}' && !isAtEnd()){
					isb.append(advance())
				}

				val innerScanner = Scanner(isb.toString(), sunlite)
				val interpolatedTokens = innerScanner.scanTokens(currentFile).dropLast(1)
				tokens.addAll(interpolatedTokens)

				tokens.add(Token(PLUS, "+", null, line, currentFile, Token.Position(lineStart,lineCurrent)))

				if(isAtEnd()){
					sunlite.error(line, "Expected '}' after string interpolated expression.")
					return
				}
				advance()
				sb.clear()
			}

			sb.append(peek())
			advance()
		}

		if (isAtEnd()) {
			sunlite.error(line, "Unterminated string.")
			return
		}

		// The closing ".
		advance()

		// Trim the surrounding quotes.
		addToken(STRING, sb.toString())
	}

	private fun number() {
		while (peek() in '0'..'9') advance()

		// Look for a fractional part.
		if (peek() == '.' && peekNext() in '0'..'9') {
			// Consume the "."
			advance()

			while (peek() in '0'..'9') advance()
		}

		addToken(
			NUMBER,
			source.substring(start, current).toDouble()
		)
	}

	private fun identifier() {
		while (isAlphaNumeric(peek())) advance()


		val text = source.substring(start, current)
		var type = keywords[text]
		if (type == null) type = IDENTIFIER
		addToken(type)

	}

	private fun isAlpha(c: Char): Boolean {
		return (c in 'a'..'z') ||
				(c in 'A'..'Z') || c == '_'
	}

	private fun isAlphaNumeric(c: Char): Boolean {
		return isAlpha(c) || c in '0'..'9'
	}


}